### PR TITLE
Add conflict support with redis

### PR DESCRIPTION
**Pull request by rystills**
Re-implementation of backend related conflict code from HaoxinLuo using redis. Front-end portion still needs to be properly re-implemented, and efficiency of conflicts code in section.rb must be assessed.

### CODE REVIEW

File: app/models/section.rb [hunk 1]
```diff
@@ -20,4 +22,26 @@ def conflicts_with(section)
     end
     false
   end
+  
+  def conflicts
+    Redis.current.smembers id
+  end
+  
+  private
+  def type_cast_periods_start_end
```
Comment (Bad-Science) [inline app/models/section.rb]: this method should not be needed, please remove. it was a bad workaround to a bug.

File: Gemfile.lock [hunk 2]
```diff
@@ -101,6 +101,7 @@ GEM
       factory_girl (~> 4.7.0)
```
Comment (Bad-Science) [inline Gemfile.lock]: remove changed to Gemfile.lock from this PR if it is coming from windows.

File: config/routes.rb [hunk 3]
```diff
@@ -9,6 +9,7 @@
       resources :courses,     only: [:index]
       resources :sections,    only: [:index]
       resources :schedules,   only: [:index]
+	  resources :conflicts,   only: [:index]
```
Comment (Bad-Science) [inline config/routes.rb]: what is this route for?

Comment (rystills) [inline config/routes.rb]: There does not appear to be a 'conflicts' request anywhere. Removed this conflicts route.

File: Gemfile.lock (hunk 2 above)
Comment (rystills) [inline Gemfile.lock]: Removed these changes to Gemfile.lock caused by the windows environment, excluding the additions of redis (as these should have been generated due to the redis gem regardless of environment).

File: app/models/section.rb (hunk 1 above)
Comment (rystills) [inline app/models/section.rb]: Removed the type_cast_periods_start_end method, as period numbers should either no longer need to be converted from strings into ints, or this functionality should be done elsewhere (as we discussed).

**Comment by copperwater** [comment]
This needs Redis pipelining added and potentially (?) more Docker setup. Moving back to WIP status.

File: app/assets/javascripts/controllers/courses/controller.js [hunk 4]
```diff
@@ -9,47 +9,192 @@ Yacs.views.courses = function (target, params) {
...
+        }), cid);
+      }
+    });
+  };
+
+  /**
+   * Given the data returned from the API, populate the conflicts cache with
+   * the data for conflicts.
+   */
+  var populateConflictsCache = function(data) {
```
Comment (Bad-Science) [inline app/assets/javascripts/controllers/courses/controller.js]: If we're fully populating this every time we load the view anyway, I don't think it makes sense to have a global cache. I think we should just use an object within the scope of the controller.

Comment (copperwater) [inline app/assets/javascripts/controllers/courses/controller.js]: Note that this isn't being fully populated; section ids that are already in the cache don't get overwritten. 

The advantage to using a conflicts cache is that parts of the application that don't use the courses API will be able to see and display the conflicts.

E.g. you select some conflicting sections, then go to the schedules page. This shows no possible schedules, but it is able to tell you what is conflicting. (This behavior doesn't exist right now.)

On the other hand, I can't see the schedule page _not_ making a call to the courses API as long as we let the user have the ability to deselect courses, and nothing else uses the conflict cache at this time, so perhaps it's fine to remove this.

Comment (copperwater) [inline app/assets/javascripts/controllers/courses/controller.js]: Another note: the whole frontend cache shouldn't be scrapped. It'll let us do things like easily show the names of conflicting courses (see #164).

**Comment by copperwater** [comment]
Closing this, I wasn't able to get ahold of @rystills to PR into his fork at the right time. I've since merged #226, which is the same thing from my fork and contains the later commits.

**Pull request closed without merging**